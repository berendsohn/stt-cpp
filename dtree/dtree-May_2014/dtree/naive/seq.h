// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_NAIVE_SEQ_H_
#define DTREE_NAIVE_SEQ_H_

#include "dtree/naive/common.h"

namespace naive {

template<typename Base>
class EndSeq : public Base {
 public:
  EndSeq() : solid_() {}
  EndSeq* solid(int i) const { return solid_[i]; }
  void set_solid(int i, EndSeq* solid_i) { solid_[i] = solid_i; }
  template<typename Forwarder>
  void CopyFrom(const EndSeq& node, Forwarder forward);
 private:
  EndSeq* solid_[2];
  DISALLOW_COPY_AND_ASSIGN(EndSeq);
};

template<typename Group, typename Base>
class WithValue : public Base {
 public:
  typedef Group Group_;
  typename Group::Type value() const { return value_; }
  void set_value(typename Group::Type value) { value_ = value; }
  void add_to_value(typename Group::Type delta) {
    set_value(Group::Plus(value(), delta));
  }
  void subtract_from_value(typename Group::Type delta) {
    set_value(Group::Minus(value(), delta));
  }
  template<typename Node>
  static typename Group::Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, typename Group::Type value);
  template<typename Node>
  static void AddToSeq(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromSeq(Node* node, typename Group::Type delta);

 protected:
  WithValue() : value_() {}

 private:
  typename Group::Type value_;
};

template<typename Semigroup, typename Base>
class WithAggr : public Base {
 public:
  typedef Semigroup Semigroup_;
  typename Semigroup::Type singleton_aggr() const {
    return Semigroup::AggrFromValue(this->value());
  }
  template<typename Node>
  static typename Semigroup::Type AggrSeq(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostSeq(int dir, Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeftmostSeq(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRightmostSeq(Node* node, const Predicate& predicate);
 protected:
  WithAggr() {}
};

template<typename Base>
class WithReverseBy : public Base {
 private:
  typedef typename Base::Group_ Group;
 public:
  template<typename Node>
  static void ReverseSeq(Node* node);
 protected:
  WithReverseBy() {}
};

template<typename Base>
class WithReverse : public Base {
 public:
  template<typename Node>
  static void ReverseSeq(Node* node);
 protected:
  WithReverse() {}
};
}  // namespace naive
#endif  // DTREE_NAIVE_SEQ_H_
