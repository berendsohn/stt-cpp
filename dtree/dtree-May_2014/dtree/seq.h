// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_SEQ_H_
#define DTREE_SEQ_H_

#include "dtree/selector.h"

#ifdef NAIVE_DTREE
#include "dtree/naive/seq.h"
#else
#include "dtree/selfadjust/seq.h"
#endif
#endif  // DTREE_SEQ_H_
