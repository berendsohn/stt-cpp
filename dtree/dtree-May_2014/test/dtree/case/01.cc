#define DTREE_CONFIG EndSeq<WithReverse<WithReverse<WithAggr<dtree::CountAndSum<double, double>, dtree::IndexByCount_<double>, WithAggr<dtree::Max<double>, dtree::GreaterEqual_<double>, WithValue<dtree::Add<double>, Begin > > > > > >
#include "test/dtree/seq.cc"
