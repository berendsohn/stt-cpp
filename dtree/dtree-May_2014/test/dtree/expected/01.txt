<<<
  3: v20
  8: v14
  14: v8
  11: v-10
  6: v11
  13: v-31
  7: v-2
  12: v11
  9: v8
  4: v11
  15: v0
  1: v4
  5: v19

  10: v0
  2: v-6
>>>
<<<
    /3: v6 a0 aC1S0 f0 f1
  @8: v14 a0 aC2S-6 f0 f0
        /13: v-21 a42 aC2S42 f1 f1
          \6: v42 a0 aC1S0 f0 f1
      /11: v-8 a21 aC4S18 f0 f1
        \14: v18 a0 aC1S0 f1 f0
    \7: v-16 a21 aC11S51 f0 f0
      \12: v13 a8 aC6S-13 f1 f1
          /9: v8 a3 aC2S3 f1 f1
            \4: v3 a0 aC1S0 f0 f0
        \15: v-11 a19 aC5S42 f1 f1
          \1: v4 a15 aC2S15 f1 f1
            \5: v15 a0 aC1S0 f0 f1

  @10: v0 a0 aC1S0 f1 f1
    \2: v-6 a0 aC1S0 f0 f0
>>>
